### DISCUSSION TITLE
Unable to mock the environment variable used to configure the root_path while testing Fast API code using pytest

### DISCUSSION THREAD

**Discussion by kunal-03** [fastapi/fastapi#8627] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
I have a simple fast api code as follow main.py

from fastapi import FastAPI
import os

app = FastAPI(root_path=f"/{os.environ['EnvName']}/")

@app.get("/getrecord")
def root():
    return {"message": "Hello World"}

Below is the pytest code test_getrecord.py

import pytest
from starlette.testclient import TestClient
from main import app

def test_dummy_api():
    monkeypatch.setenv("EnvName", "dummy")
    client=TestClient(app)
    response = client.get("/getrecord")
    assert response.status_code == 200
```

### Description

when running the above pytest code pytest -v ./test/test_getrecord.py getting below error:

(FaSTAPIApplication) C:\Users\abc\Documents\FastAPIWork\FastAPIHttpWrapper>pytest -v ./test/test_getrecord.py
============================================================ test session starts ============================================================
platform win32 -- Python 3.7.11, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- c:\users\abc\anaconda3\envs\fastapiapplication\python.exe   
cachedir: .pytest_cache
rootdir: C:\Users\abc\Documents\FastAPIWork\FastAPIHttpWrapper
collected 0 items / 1 error

================================================================== ERRORS ===================================================================
__________________________________________________ ERROR collecting test/test_getrecord.py __________________________________________________ 
test\test_getrecord.py:5: in <module>
    from main import app
main.py:4: in <module>
    app = FastAPI(root_path=f"/{os.environ['EnvName']}/")
..\..\..\anaconda3\envs\fastapiapplication\lib\os.py:681: in __getitem__
    raise KeyError(key) from None
E   KeyError: 'EnvName'
========================================================== short test summary info ========================================================== 
ERROR test/test_getrecord.py - KeyError: 'EnvName'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

### Operating System

Windows

### FastAPI Version

0.68.1

### Python Version

Python 3.7.11

### Additional Context

Passing root_path dynamically because the value is changing based on environment, thus that needs to be dynamic, but unable to execute the test cases because not able to mock the os environment, please let me know how can I move forward with this.

**Comment by DanielElisenberg** [comment, marked as answer]
Hey, Kunal 👋🏻 

Your issue stems from importing the app before you mokeypatch the environment.
```py
# You import the app first
from main import app
# The app then attempts to do this:
app = FastAPI(root_path=f"/{os.environ['EnvName']}/")
# This does not work as you have not yet reached the monkeypatch statement
monkeypatch.setenv("EnvName", "dummy")
```

Have a look into [pytest-dotenv](https://github.com/quiqua/pytest-dotenv). This tool monkeypatches your environment from an .env file before any of your test code runs.
